### ISSUE TITLE
`brew audit --cask --signing` seems not working on certain devices

### ISSUE THREAD

**Issue by WangEdward** [Homebrew/brew#15476]
### `brew doctor` output

```shell
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.0.19-9-g8aaf99e
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8aaf99e1d70c5f50a828c34c3219ecfb17203ba0
Last commit: 15 hours ago
Core tap JSON: 23 May 12:29 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.40.1 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: false
```

### What were you trying to do (and why)?

When running `brew audit --cask --signing` on my local device, [Signature verification Error](https://github.com/Homebrew/homebrew-cask/actions/runs/5057504644/jobs/9076372311) will not be presented.

### What happened (include all command output)?

<details>
    <summary>brew audit --cask --online --new-cask --signing --arch=arm -vv --debug --strict wangedward/testcask/youtube-music</summary>

```
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromTapLoader): loading wangedward/testcask/youtube-music
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromTapLoader): loading wangedward/testcask/youtube-music
/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/bin/bundle clean
/usr/bin/env XDG_CACHE_HOME=/Users/edward/Library/Caches/Homebrew/style /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby -W1 --disable=gems,rubyopt -- /opt/homebrew/Library/Homebrew/utils/rubocop.rb --format json --force-exclusion --parallel --extra-details --config /opt/homebrew/Library/.rubocop.yml /opt/homebrew/Library/Taps/wangedward/homebrew-testcask/Casks/youtube-music.rb
==> Auditing Cask youtube-music on os ventura and arch arm
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromTapPathLoader): loading /opt/homebrew/Library/Taps/wangedward/homebrew-testcask/Casks/youtube-music.rb
==> Auditing version :latest does not appear as a string ('latest')
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-9-g8aaf99e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --location https://api.github.com/repos/th-ch/youtube-music/releases/latest --header Accept:\ application/vnd.github\+json --write-out '
'\%\{http_code\} --header Authorization:\ token\ ****** --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20230523-32759-rjxlyc
==> Auditing signing
==> Downloading https://github.com/th-ch/youtube-music/releases/download/v1.20.0/YouTube-Music-1.20.0-arm64.dmg
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-9-g8aaf99e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --fail --location --silent --head https://github.com/th-ch/youtube-music/releases/download/v1.20.0/YouTube-Music-1.20.0-arm64.dmg
Already downloaded: /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
==> Checking quarantine support
/usr/bin/env /usr/bin/xattr -h
/usr/bin/env /usr/bin/swift -target arm64-apple-macosx13 /opt/homebrew/Library/Homebrew/cask/utils/quarantine.swift
==> Quarantine is available.
==> Verifying Gatekeeper status of /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
/usr/bin/env /usr/bin/xattr -p com.apple.quarantine /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
==> /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg is quarantined
==> Verifying checksum for 'f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg'
/usr/bin/env hdiutil imageinfo -format /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
/usr/bin/env hdiutil attach -plist -nobrowse -readonly -mountrandom /private/tmp/d20230523-32759-18lwusn /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
/usr/bin/env find . -print0
/usr/bin/env mkbom -s -i /private/tmp/20230523-32759-9jhy9k.list -- /private/tmp/20230523-32759-1mldmfo.bom
/usr/bin/env ditto --bom /private/tmp/20230523-32759-1mldmfo.bom -- /private/tmp/d20230523-32759-18lwusn/dmg.Py2ZcZ /private/tmp/d20230523-32759-1kbvm5n
/usr/bin/env diskutil info -plist /private/tmp/d20230523-32759-18lwusn/dmg.Py2ZcZ
/usr/bin/env diskutil eject /private/tmp/d20230523-32759-18lwusn/dmg.Py2ZcZ
/usr/bin/env cp -pR /private/tmp/d20230523-32759-1kbvm5n/YouTube\ Music.app/. /private/tmp/d20230523-32759-1ukqxwh/YouTube\ Music.app
/usr/bin/env spctl --assess --type install /private/tmp/d20230523-32759-1ukqxwh/YouTube\ Music.app
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-9-g8aaf99e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --location https://api.github.com/repos/th-ch/youtube-music/releases/latest --header Accept:\ application/vnd.github\+json --write-out '
'\%\{http_code\} --header Authorization:\ token\ ****** --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20230523-32759-7nalwj
==> Auditing URL format
==> Auditing GitHub prerelease
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-9-g8aaf99e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --location https://api.github.com/repos/th-ch/youtube-music/releases/tags/v1.20.0 --header Accept:\ application/vnd.github\+json --write-out '
'\%\{http_code\} --header Authorization:\ token\ ****** --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20230523-32759-j4yqzu
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-9-g8aaf99e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --connect-timeout 15 --max-time 25 --retry 3 --retry-max-time 25 --dump-header - --output /private/tmp/20230523-32759-11ja69e --location https://github.com/th-ch/youtube-music/releases/download/v1.20.0/YouTube-Music-1.20.0-arm64.dmg
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Mozilla/5.0\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10_12_3\)\ AppleWebKit/602.4.8\ \(KHTML,\ like\ Gecko\)\ Version/10.0.3\ Safari/602.4.8 --header Accept-Language:\ en --connect-timeout 15 --max-time 25 --retry 3 --retry-max-time 25 --dump-header - --output /private/tmp/20230523-32759-i7yqc6 --location https://th-ch.github.io/youtube-music/
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-9-g8aaf99e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --location https://api.github.com/repos/th-ch/youtube-music --header Accept:\ application/vnd.github\+json --write-out '
'\%\{http_code\} --header Authorization:\ token\ ****** --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20230523-32759-rmj5d2
==> Auditing GitHub repo
==> Downloading https://github.com/th-ch/youtube-music/releases/download/v1.20.0/YouTube-Music-1.20.0-arm64.dmg
Already downloaded: /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
==> Verifying Gatekeeper status of /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
/usr/bin/env /usr/bin/xattr -p com.apple.quarantine /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
==> /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg is quarantined
==> Verifying checksum for 'f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg'
==> Auditing pkg stanza: allow_untrusted
==> Auditing stanzas which require an uninstall
==> Auditing single uninstall_* and zap stanzas
==> Auditing required stanzas
==> Auditing preflight and postflight stanzas
==> Auditing sha256 :no_check with version :latest
==> Auditing sha256 string is a legal SHA-256 digest
==> Auditing sha256 is not a known invalid value
```

</details>

<details>
    <summary>brew audit --cask --online --new-cask --signing --arch=arm wangedward/testcask/youtube-music</summary>

```
==> Downloading https://github.com/th-ch/youtube-music/releases/download/v1.20.0/YouTube-Music-1.20.0-arm64.dmg
Already downloaded: /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
==> Downloading https://github.com/th-ch/youtube-music/releases/download/v1.20.0/YouTube-Music-1.20.0-arm64.dmg
Already downloaded: /Users/edward/Library/Caches/Homebrew/downloads/f84b30e206741a88aa7d5eacaaf13e2d6f43f65a40ceed3df8ea0d89b7e1a3b3--YouTube-Music-1.20.0-arm64.dmg
```

</details>

### What did you expect to happen?

As displayed in the [workflow](https://github.com/Homebrew/homebrew-cask/actions/runs/5057504644/jobs/9076372311), there should be a 'Signature verification failed' notice. (Maybe related https://github.com/Homebrew/homebrew-cask/pull/147465)

**Comment by WangEdward** [comment]
Is the signing check being bypassed by adding a caveat? (checks pass at [my fork's workflow](https://github.com/WangEdward/homebrew-cask/actions/runs/5057980837/jobs/9077527344?pr=1))

**Comment by reitermarkus** [comment]
> Is the signing check being bypassed by adding a caveat?

There is no condition in the code that checks for this caveat.

Your command debug output shows `==> Auditing signing`, which indicates that the signing check is in fact not skipped. Maybe a condition inside the signing check if wrongfully triggered.

**Comment by WangEdward** [comment]
Thanks for your reply. I think I found the reason why my local machine's signing check is not working:

**In the case that `spctl --status` is `assessments enabled`**:

``` bash
the output for a signed app:
➜ spctl --assess --type install '/Applications/Google Chrome.app'
[nothing returned]

➜ spctl --assess --type install '/Applications/Google Chrome.app' -v
/Applications/Google Chrome.app: accepted
source=Notarized Developer ID

the output for a unsigned app:
➜ spctl --assess --type install '/Applications/YouTube Music.app'          
/Applications/YouTube Music.app: code has no resources but signature indicates they must be present
```

However, some users may follow the guide on the internet to enable 'Allow applications downloaded from Anywhere', they might run the command `sudo spctl --master-disable`, which disables Gatekeeper at the same time.

**In this situation, `spctl --status` is `assessments disabled`**:
```bash
the output for a signed app:
➜ spctl --assess --type install '/Applications/Google Chrome.app'
[nothing returned]

➜ spctl --assess --type install '/Applications/Google Chrome.app' -v
/Applications/Google Chrome.app: accepted
source=Notarized Developer ID
override=security disabled

the output for a unsigned app:
➜ spctl --assess --type install '/Applications/YouTube Music.app'
[nothing returned] -> (the same as when it's signed)

➜ spctl --assess --type install '/Applications/YouTube Music.app' -v
/Applications/YouTube Music.app: accepted
override=security disabled
```
(`spctl --assess --type install ` command is according to: https://github.com/Homebrew/brew/blob/129f434894bc1b0a492075c57c9d00cf0bc74536/Library/Homebrew/cask/audit.rb#L494 

In conclusion, the check will not work when Gatekeeper is disabled, maybe add an additional warning in this situation when checking the signature?

But at the same time, this still doesn't explain why https://github.com/WangEdward/homebrew-cask/actions/runs/5057980837/jobs/9077527344?pr=1 can pass the signature check, as the github workflow should have the Gatekeeper enabled by default.
